### PR TITLE
#162163263 Users should receive descriptive error messages on unsuccessful registration

### PR DESCRIPTION
**Pull request by kafuuma**
#### What does this PR do?
This PR delivers a feature for users to receive descriptive error messages on successful registration

#### Description of the tasks to be completed?
- User password must be between 8 to 128 characters and, passwords must include at least a number and a capital letter
- Usernames should be between 3 to 8 characters, names should only contain letters, numbers, and underscores.
- A user cannot register with an existing username or email address
- Email addresses must be valid

#### How should this be manually tested?
- Paste URL in Postman
`https://ah-backend-athena-staging-pr-9.herokuapp.com/api/users/`
```
Register with valid credentials as described above
Try also to register with invalid credentials
```

Screenshots 
1. Registering with short username
image
2. Registering with Long username
image
3. Invalid password
image

#### What are the relevant pivotal tracker stories
[#162163263](https://www.pivotaltracker.com/story/show/162163263)

### CODE REVIEW

**Review by bibangamba — APPROVED** [review]
Hi Henry, couldn't really test the PR with the link provided. It points to a feature-branch deploy that is unrelated (https://ch-heroku-deploy-162163258.herokuapp.com/api/users/). For this PR, Heroku created a review app (https://ah-backend-athena-staging-pr-9.herokuapp.com) which I think you should refer to in the testing section. Then if we could get more details on testing with screenshots (for this case);
something like
1. create a POST request on Postman using the review app URL for this PR, append '/api/users'
2. submit a JSON body with invalid data e.g. username is 2 characters, to see the error raised.
etc
image

Otherwise, this is nice work! I've tested it using the review app URL attached to the PR.

**Comment by kafuuma** [comment]
Alright, Andrew,
Let me include the right link plus snapshots

**Comment by KicaRonaldOkello** [comment]
You have a typo on your PR. it is supposed to be **unsuccessful registration.**
screen shot 2018-12-05 at 13 07 30

**Review by Nta1e — APPROVED** [review]

File: authors/apps/authentication/test/test_validated_users.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    def test_user_registration_weak_password(self):
+        user = self.generate_user(self.username, self.email, 'weakpassword')
+        response = self.client.post('/api/users/', user, format='json')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_user_registration_username_long(self):
+        username = 'x'*300
+        user = self.generate_user(username, self.email, self.password)
+        response = self.client.post('/api/users/', user, format='json')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (Kasulejoseph) [inline authors/apps/authentication/test/test_validated_users.py:64]: Thank for the tests @kafuuma, could you  PLEASE assert for return messages as well ..

**Review by Kasulejoseph — CHANGES_REQUESTED** [review]

**Review by Kasulejoseph — APPROVED** [review]

**Review by joshtrigger — APPROVED** [review]

**Review by KicaRonaldOkello — APPROVED** [review]

**Review by bibangamba — APPROVED** [review]

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -8,14 +8,13 @@
...
 class UserManager(BaseUserManager):
-    """
-    Django requires that custom users define their own Manager class. By
-    inheriting from `BaseUserManager`, we get a lot of the same code used by
-    Django to create a `User` for free. 
 
-    All we have to do is override the `create_user` function which we will use
-    to create `User` objects.
-    """
+    # Django requires that custom users define their own Manager class. By
```
Comment (collin5) [inline authors/apps/authentication/models.py]: Use doc string for this

File: authors/apps/authentication/models.py [hunk 3]
```diff
@@ -101,19 +99,19 @@ def __str__(self):
 
     @property
     def get_full_name(self):
-        """
-        This method is required by Django for things like handling emails.
-        Typically, this would be the user's first and last name. Since we do
-        not store the user's real name, we return their username instead.
-        """
+
```
Comment (collin5) [inline authors/apps/authentication/models.py]: Same as above

File: authors/apps/authentication/models.py [hunk 4]
```diff
@@ -101,19 +99,19 @@ def __str__(self):
...
         return self.username
 
     def get_short_name(self):
-        """
-        This method is required by Django for things like handling emails.
-        Typically, this would be the user's first name. Since we do not store
-        the user's real name, we return their username instead.
-        """
+
+        # This method is required by Django for things like handling emails.
```
Comment (collin5) [inline authors/apps/authentication/models.py]: ☝️

File: authors/apps/authentication/serializers.py [hunk 5]
```diff
@@ -1,28 +1,74 @@
...
         # or response, including fields specified explicitly above.
+
         fields = ['email', 'username', 'password']
 
+    def validate_password(self, password):
+        """
+        User's password should be between 8 to 128 characters
+        and  should contain atleas one number and a capital letter
+        """
+        if re.match('^.{8,128}$', password) is None:
```
Comment (collin5) [inline authors/apps/authentication/serializers.py]: Use `len()` for this. It's cheaper than using regex.

File: authors/apps/authentication/serializers.py [hunk 6]
```diff
@@ -1,28 +1,74 @@
...
+        return email
+
+    def validate_username(self, username):
+        username_db = User.objects.filter(username=username)
+        if username_db.exists():
+            raise serializers.ValidationError(
+                'The user name you entered is already taken, try another one'
+            )
+
+        if re.match('^.{3,10}$', username) is None:
```
Comment (collin5) [inline authors/apps/authentication/serializers.py]: Same as above. Use `len()`

File: authors/apps/authentication/serializers.py [hunk 7]
```diff
@@ -57,17 +103,27 @@ def validate(self, data):
...
         # model, we set `USERNAME_FIELD` as `email`.
+
         user = authenticate(username=email, password=password)
 
         # If no user was found matching this email/password combination then
         # `authenticate` will return `None`. Raise an exception in this case.
         if user is None:
             raise serializers.ValidationError(
-                'A user with this email and password was not found.'
+                'You typed a wrong password, please try again'
```
Comment (collin5) [inline authors/apps/authentication/serializers.py]: Any reasons why you changed this message? Seems better as it is to me.

File: authors/apps/authentication/test/test_authentication.py [hunk 8]
```diff
@@ -44,82 +47,3 @@ def test_user_registration_wrong_email_format(self):
         user = self.generate_user('athena', 'athenmail', 'password@user')
         response = self.client.post('/api/users/', user, format='json')
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-
-    def test_user_login(self):
```
Comment (collin5) [inline authors/apps/authentication/test/test_authentication.py]: Any reasons why you completely removed these tests?

**Review by collin5 — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/models.py (hunk 2 above)
Comment (kafuuma) [reply authors/apps/authentication/models.py]: The hash comments were already there in the codebase. Actually, I tried to change them to doc strings and got merge conflicts after Kasule's PR was merged
So I reverted them, that's why they are in red

File: authors/apps/authentication/serializers.py (hunk 5 above)
Comment (kafuuma) [reply authors/apps/authentication/serializers.py]: Let me work on that

File: authors/apps/authentication/serializers.py (hunk 6 above)
Comment (kafuuma) [reply authors/apps/authentication/serializers.py]: This too

File: authors/apps/authentication/serializers.py (hunk 7 above)
Comment (kafuuma) [reply authors/apps/authentication/serializers.py]: This message wasn't descriptive enough to me, at first, a user could not know which of the two between email or password that were typed is wrong. So I refactored the code so that I specifically notify the user, which of the two wasn't typed correctly. First I first check the email alone to find out if the user exists, then I proceed to authenticate to check whether the user typed the password correctly. There I can clearly notify the user what specifically went wrong

File: authors/apps/authentication/test/test_authentication.py (hunk 8 above)
Comment (kafuuma) [reply authors/apps/authentication/test/test_authentication.py]: I discovered that Kasule had also written the same tests while working on the authentication with jwt feature, so it was like duplicate code

Comment (collin5) [reply authors/apps/authentication/test/test_authentication.py]: But diff shows you deleted them. Meaning they were already part of the version you branched off (I'm assuming you branched of `develop`). You should therefore leave them as they are or refactor if necessary.

File: authors/apps/authentication/serializers.py (hunk 7 above)
Comment (collin5) [reply authors/apps/authentication/serializers.py]: I think it's better to let this message be generic (leave as is). Otherwise, this can allow user enumeration.

Comment (kafuuma) [reply authors/apps/authentication/serializers.py]: alright

File: authors/apps/authentication/test/test_authentication.py (hunk 8 above)
Comment (kafuuma) [reply authors/apps/authentication/test/test_authentication.py]: Ya I branched of develop and created them and then deleted them after
The code base has no doc strings. But I can put them pack hope ti will not cause conflicts again

Comment (collin5) [reply authors/apps/authentication/test/test_authentication.py]: Same here, just rebase against `develop` and resolve the conflicts before pushing.

File: authors/apps/authentication/models.py (hunk 2 above)
Comment (collin5) [reply authors/apps/authentication/models.py]: You could just rebase against `develop` and resolve the conflicts before pushing.

File: authors/apps/authentication/test/test_authentication.py (hunk 8 above)
Comment (kafuuma) [reply authors/apps/authentication/test/test_authentication.py]: alright

**Review by collin5 — APPROVED** [review]
